src/Microsoft.OData.Core/ObjectPool.cs

```diff
@@ -0,0 +1,91 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="ObjectPool.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+
+namespace Microsoft.OData
+{
+    /// <summary>
+    /// A pool of objects.
+    /// </summary>
+    /// <typeparam name="T">The type of objects to pool.</typeparam>
+    /// <remarks>This object pool is not thread-safe and not intended to be shared across concurrent requests.
+    /// Consider using the DefaultObjectPool in 8.x once we drop support for netstandard 1.1
+    /// </remarks>
+    internal class ObjectPool<T>
+    {
+        private readonly Func<T> objectGenerator;
+        private readonly ObjectWrapper[] items;
+        private T firstItem;
+
+        /// <summary>
+        /// To initialize the object pool.
+        /// </summary>
+        /// <param name="objectGenerator">Used to create an instance of a <typeparamref name="T"/>.</param>
+        public ObjectPool(Func<T> objectGenerator)
+        {
+            this.objectGenerator = objectGenerator ?? throw new ArgumentNullException(nameof(objectGenerator));
+            items = new ObjectWrapper[32];
+        }
+
+        /// <summary>
+        /// Gets an object from the pool if one is available, otherwise creates one.
+        /// </summary>
+        /// <returns>A <typeparamref name="T"/>.</returns>
+        public T Get()
+        {
+            T item = firstItem;
+
+            if (item != null)
+            {
+                firstItem = default(T);
+                return item;
+            }
+

```

Comment:
#Resolved

Reply:
I don't know what was here before, but there's really no need to have an instance of `ObjectPool` with a `null` value for `items`. I don't see us ever set it back to `null`, so we could mark it `readonly` if we initialize it in the constructor. Also, doing this means that we will check the conditional every time `Get` is called, which seems suboptimal.

Reply:
Good point about calling `if (items == null)` on every `Get()` call seeming suboptimal. My suggestion was based on eliminating an array allocation (and deallocation) when it's not necessary. I would assume it would be much easier for JIT or CPU to optimize `if (items == null)` away via branch prediction or some other techniques if `items == null` evaluates to the same value most of the times. So in both cases there's trade-off to be made. But this lazy initialization pattern seems common in BCL source code, e.g.:
- https://source.dot.net/#System.Text.Json/System/Text/Json/Serialization/WriteStack.cs,95
- https://source.dot.net/#System.Text.Json/System/Text/Json/BitStack.cs,61
- https://source.dot.net/#System.Private.CoreLib/Dictionary.cs,506

Can this be evaluated using benchmarks?